kernel/proc: add scheduler description (#132)
phoenix-rtos/phoenix-rtos-doc: kernel/proc/scheduler.md

# Kernel - Processes and threads - Scheduler
The operating system scheduler controls threads execution with a predetermined policy. It is a part of Phoenix-RTOS having the most significant influence on the performance and responsiveness of the whole system.

## Scheduling policy

> The scheduling algorithm is defined in the `threads_schedule` function, which is invoked by a timer hardware interrupt. Each function's invocation changes the context of the currently executed thread to the next one, available in a thread's ready queue. The scheduling policy determined which thread shall be chosen for the next execution. In Phoenix-RTOS, there is used a priority preempted round-robin algorithm. There are eight priority levels, the smallest value holds the highest priority.

Reviewer: It is invoked by timer interrupt or voluntary reschedule (`hal_cpuReschedule`)

The thread management unit contains an eight-element array, where each of them holds a pointer to the list of threads of the same priority. A scheduling algorithm is defined as follows:

> 1. The spinlock is set before any operations on common thread management data.

Reviewer: What spinlock exactly? It does matter (on multi-core systems), so it's good to emphasize this

> 2. The unit holds information about the running threads of each processor's core. The current thread's context for the interrupted core is saved and added to the end of the priority list.

Reviewer: First sentence is imho redundant. "priority list" is imho misleading - it's a name of data structure we don't have and don't use.
```suggestion
2. The current thread's context for the interrupted core is saved and added to the end of the priority list.
```

> 3. The next available thread with the highest priority is selected to be run next and is removed from the ready thread's list. If a selected thread is a ghost, it is added to the ghosts list and the reaper thread is waked up.

Reviewer: Repetition. We need to clarify what a ghost is (it's not a common concept). Perhaps some reference to a place in documentation that explains this?
```suggestion
3. The next available thread with the highest priority is selected to be run and is removed from the ready thread's list. If a selected thread is a ghost, it is added to the ghosts list and the reaper thread is waked up.
```

4. For the selected thread, the following actions are performed:
    * a pointer to the current thread is changed to the selected one,

>     * a kernel stack is changed to the stack of a new thread,

Reviewer: It's not, it would crash the system
```suggestion
    * a pointer to the kernel stack is updated to the stack of a new thread,
```

    * a memory map is changed to the map associated with the thread's process,
    * signal handlers are performed,
    * performance data is saved in a perf unit,
5. The cpu usage is updated for the current and selected thread.

> 6. At the end of the modification of the common thread management data, the spinlock is cleared.

Reviewer: There is no information about context switch - how and where it's done?

Pointer to the stack in context of current thread is updated with a pointer to the stack of selected thread. When we exit scheduler context restore is performed - thanks to this change in saved SP we restore selected thread instead.

# Kernel - Processes and threads - Scheduler
The operating system scheduler controls threads execution with a predetermined policy. It is a part of Phoenix-RTOS having the most significant influence on the performance and responsiveness of the whole system.

## Scheduling policy
The scheduling algorithm is defined in the `threads_schedule` function. It is invoked by timer interrupt or voluntary reschedule (`hal_cpuReschedule`). Each function's invocation changes the context of the currently executed thread to the next one, available in a thread's ready queue. The scheduling policy determined which thread shall be chosen for the next execution. In Phoenix-RTOS, there is used a priority preempted round-robin algorithm. There are eight priority levels, the smallest value holds the highest priority.

The thread management unit contains an eight-element array, where each of them holds a pointer to the list of threads of the same priority. A scheduling algorithm is defined as follows:

> 1. The spinlock within thread management unit is set before any operations on common data.

Reviewer: Maybe just say `threads_common.spinlock` is set? I don't know how technical we want to be in this documentation, there're already references to functions, so I guess it would be ok

> 2. The current thread's context for the interrupted core is saved and added to the end of the priority list.

Reviewer: I think this clarifies that we talk about one of 8 lists, not a priority list.
```suggestion
2. The current thread's context for the interrupted core is saved and added to the end of its priority list.
```

> 3. The next available thread with the highest priority is selected to be run and is removed from the ready thread's list. If a selected thread is a ghost (a thread whose process has ended execution) and has not been executed in a supervisor mode, it is added to the ghosts list and the reaper thread is waked up.

Reviewer: ```suggestion
3. The next available thread with the highest priority is selected to be run and is removed from the ready threads list. If a selected thread is a ghost (a thread whose process has ended execution) and has not been executed in a supervisor mode, it is added to the ghosts list and the reaper thread is woke up.
```

4. For the selected thread, the following actions are performed:
    * a global pointer to the current thread is changed to the selected one,
    * a pointer to the kernel stack is updated to the stack of a new thread,
    * a memory map is changed to the map associated with the thread's process,
    * signal handlers are performed,
    * performance data is saved in a perf unit,

>     * in the `hal_cpuRestore` pointer to the stack in a context of current thread is updated with a pointer to the stack of selected thread. When the scheduler finishes work, the context restore is performed. The stack of selected thread is restored from the current thread which was previously saved by the `hal_cpuRestore` function.

Reviewer: tbh I don't really understand last sentence, it seem somewhat redundant to me, I propose to add information what context is restored to the previous sentence.
```suggestion
    * in the `hal_cpuRestore` pointer to the stack in a context of current thread is updated with a pointer to the stack of selected thread. When the scheduler finishes work, the context of selected thread restore is performed.
```
